### DISCUSSION TITLE
Call function when a socket gets garbage collected?

### DISCUSSION THREAD

**Discussion by drawnwren** [miguelgrinberg/python-socketio#1391] [General]
When `monitor_clients` is set to `True` (default), the docs say that a background monitoring task will close idle connections. Is there a way in the library for me to add a custom handler to be called when an idle connection is closed? I would like to clean up parts of my code that can't be attached to socket state directly.

**Comment by miguelgrinberg** [comment]
Yes, you can do this in the disconnect handler.

The monitoring is not for idle connections, by the way. It is for clients that stop responding to pings from the server.

**Reply by drawnwren** [reply]
Thanks for being so responsive on the discussions here. Is there a reason that a client could disconnect and the disconnect handler not fire? I've written some memory cleanup in the disconnect handler and it doesn't trigger under load.

**Reply by miguelgrinberg** [reply]
How are you determining that a client disconnected?

**Reply by drawnwren** [reply]
Hmm, ok that's fair. Here's what I observe: 

1. I allocate some state (redis pubsub subcription) in connect. 
2. I have written code to delete that same state in the `disconnect` handler
3. I'm observing that with large numbers of connections that create state, a group of servers behind an aws load balancer never cleans up the memory (I removed this specific part of the code and server no longer leaks memory)

My current guess it that the aws sticky connections cookie (tells the load balancer to route the traffic to the same server) is expiring and the clients timeout on their respective servers but the state is never cleaned.

**Reply by miguelgrinberg** [reply]
My suggestion is that you enable detailed logs for Socket.IO so that you can trace the life of a connection directly. The problem with your observations is that they are indirect. You see that there is some stuff that never goes away, but we don't know if this is because the disconnect handler did not execute or maybe it did execute but the code that releases these resources is faulty and maybe crashed. With the logs we will be able to know for sure what happened to each connection. See the troubleshooting section of the documentation if you don't know how to enable logging.

Also as a side note, your theory that the load balancer cookies are expiring is very easy to confirm or discard. Have you looked at these cookies to see how long they are configured to last? Is that duration long enough or too short?

**Reply by drawnwren** [reply]
Yes, the load balancer cookies expire after 10 minutes. So that part of the theory is easy to confirm. We've been running socketio logs, I'll write a script to check if each connection is paired with a disconnect event.
